### PR TITLE
Building inside docker container

### PR DESCRIPTION
**Pull request by vmilosevic**
Adding 
- Dockerfile
- workflow for building docker image
- project build using docker image

The Dockerfile describes how to create an image for building the tt-mlir project file. It starts with a supported base image (Ubuntu 20.04) and installs the necessary packages. The purpose of the Docker build is to:

- Define and install build dependencies
- Prepare the toolchain

During the Docker build, the project is built and tests are run to ensure that everything is set up correctly. If any dependencies are missing, the Docker build will fail.

This process also prepopulates caches for Python packages and the ccache cache in the image, which should make subsequent builds faster.

Closes: #273

### CODE REVIEW

**Comment by vmilosevic** [comment]
Maybe we can go with docker and non-docker builds in parallel for a while until we see it is stable. Also, we need non docker build if we want to verify build on mac-os
I was seeing some linter errors so I set them to "continue-on-error" but we should check whats going on before merging

**Comment by nsmithtt** [comment]
> Maybe we can go with docker and non-docker builds in parallel for a while until we see it is stable. Also, we need non docker build if we want to verify build on mac-os I was seeing some linter errors so I set them to "continue-on-error" but we should check whats going on before merging

@vmilosevic, I'm OK with replacing our bare-metal ubuntu build with the docker version.  We can keep the Mac build as is.

@vmilosevic, don't hate me, but I think we need to upgrade to ubuntu 22.04 asap.  I pinged you in the tt-mlir-ci channel with more info, but we need to upgrade metal to unblock compiler development and we need ubuntu 22.04 to upgrade metal.  I know this seems counter to what metal docs/CI says, more details in slack if you're interested.

With this in mind, maybe we want to just hold off and push our first docker container with metal upgrade all at once?  That way folks can adopt the new docker workflow along with new metal all at once.

**Review by tapspatel — APPROVED** [review]

**Comment by tapspatel** [comment]
looks awesome!

**Review by nsmithtt — APPROVED** [review]
Thank you for the quick turnaround on 22.04, this looks great!

File: .github/Dockerfile
Comment (nsmithtt) [inline .github/Dockerfile:46]: Stale comment? We can fix it in a follow on PR, not blocking

**Pull request merged**
